### PR TITLE
Added left and right mouse button handlers to MacOs Canvas widget

### PR DESCRIPTION
**Pull request by Bobw147**
toga_cocoa/widgets/canvas.py:
Added mouse event handlers for left and right mouse buttons in TogaCanvas class.
Added set_on_xxxxx to Canvas Widget class

toga/widgets/canvas.py
Added mouse event getter and setter properties with supporting documentation

core/widgets/test_canvas.py
Added unit tests for the mouse button event handlers

This change permits mouse button event handlers to be added to a canvas 

## PR Checklist:

- [x ] All new features have been tested
- [x ] All new features have been documented
- [x ] I have read the **CONTRIBUTING.md** file
- [x ] I will abide by the code of conduct

### CODE REVIEW

**Comment by Bobw147** [comment]
I have now installed better support into the virtual environment I used for the changes so will sort out the pydocstyle issues this weekend. As for naming as far as I am aware the use of the term ‘mouse’ is pretty generic because of the historical associations but am willing to be proved wrong on this. Off the top of my head I can’t think of any better naming convention but will give it more thought. I am not at all familiar with GTK so can’t really comment on that but will have a look at how to tackle the ‘not yet implemented’ feature. When I have made any necessary changes do I need to issue a new pull request? (I am pretty new to git, I come mainly from a SubVersion background for my sins in a previous life)

**Comment by freakboy3742** [comment]
No need to submit a new PR; you can add a commit to the end of the branch you're on, push the update, and this PR will be updated.

I agree that a lot of APIs have adopted "mouse" as a generic term; however, that feels a bit like a an option you adopt when you have historical compatibility to consider. We're starting from scratch, so I'd like to give some thought to what naming you'd pick if you *didn't* have historical considerations. Any thoughts on the "press / release / drag" option I suggested?

**Comment by Bobw147** [comment]
I am happy tp go with press / release/ drag. ‘Press’ is consistent with the button widget behaviour. Hopefully I will have time to look at it tomorrow, sort out the GTK+ version as  ’not implemented’ and resubmit it.

Bob

**Comment by Bobw147** [comment]
I got a bit trigger happy earlier but the most recent update should cover everything except the Gtk stubs. I am not at all familiar with Gtk so have no real idea how to code them up. Is it as simple as creating a 'set_on_x' handler with the standard 'not yet implemented' message in for each event or would it need more than that?

File: src/core/toga/widgets/canvas.py [hunk 1]
```diff
@@ -518,6 +518,12 @@ def __init__(self, id=None, style=None, on_resize=None, factory=None):
 
         # Set all the properties
         self.on_resize = on_resize
+        self.on_press = None
```
[the remark this answers is not in the document]
Comment (Bobw147) [reply src/core/toga/widgets/canvas.py]: New handlers now added to the widget constructor. Also renamed on_dragged and on right_dragged to on_drag and on_right_drag respectively

File: src/cocoa/toga_cocoa/widgets/canvas.py [hunk 2]
```diff
@@ -239,4 +281,29 @@ def rehint(self):
         self.interface.intrinsic.width = fitting_size.width
 
     def set_on_resize(self, handler):
+        """No special handling required."""
+        pass
+
+    def set_on_mouse_down(self, handler):
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/widgets/canvas.py:287]: These method names need to be updated

**Comment by Bobw147** [comment]
Hi,

Thanks for the feedback. I am keen to get this update finished so will turn it round as quickly as I can. It would be useful to know what settings you use on the linter (if any) so that I can set mine up the same (I am currently using Flake8). It is easy to miss a trailing whitespace in the plethora of docstring related messages and very annoying when I upload it only to find that it fails a test.

I will have a look at the Winforms implementation.

Regards
Bob

**Comment by Bobw147** [comment]
Ok...PyCodeStyle is blowing my mind. It objects if the line is too long. If I use continuation lines it complains that it is under-indented and adding one more tab it complains that it is over-indented. This is one of the reasons I moved to Flake8.

**Comment by freakboy3742** [comment]
@Bobw147 The tool is badly named; under the hood, it *is* flake8. Once upon a time it was pycodestyle, and the external name hasn't been updated. 

The long term plan is to fix flake8 errors in the entire codebase and then "just use flake8" - or possibly adopt black (if I can bring myself to terms with some of the outstanding bugs in black).

**Comment by Bobw147** [comment]
@freakboy3742 . I have figured out what the linter was actually trying to complain about and so now it passes all the checks. Does anything else need doing or can we call this one a wrap and get it into the next dev release?

**Comment by Bobw147** [comment]
Yes, i realised about the master issue yesterday. I am afraid that was initially down to being unfamiliar with how git works. I was under the mistaken impression that it had to come from the master. Sorry about that. I will make sure that the next changes come from a branch.

Bob

**Comment by Bobw147** [comment]
Hi Russell,

Yes, I realised about the mater issue yesterday after I had submitted the latest update. I am afraid that was down to being unfamiliar with how git worked and the linkages it uses in this scenario.  Apologies for that extra bit of complication. I will make sure that all future changes requests (and I have some in the pipeline) come from a branch.

On the matter about naming conventions what would you use? I did look at the Apple documentation after your previous request for a change of name and yes, they also use ‘right’ as the convention and also have support for a third button called ‘other’. My personal feeling is that people are familiar with the concept of 'right click’ or am I just showing my age here?

Regards
Bob

**Comment by freakboy3742** [comment]
If you check the PR that I merged (#921), you'll see I've gone with `alt` as an alternative to `right`. And yes - you're correct that "right click" is very much in the vernacular; it's also baked into at least the Cocoa APIs for this feature. However, as I said previously in the thread, we're starting with a clean slate, and as a result "it's always been done this way" isn't *necessarily* a good enough reason for retaining naming. 

As an example why this is desirable: My father is left handed, and uses a mouse in his left hand with left and right buttons switched. So, his "right" mouse events come from the "left" mouse button. So... does would he attach handlers to "on_press" or "on_right_press"?

**Pull request closed without merging**